### PR TITLE
gnome3.20: init at 3.20.0

### PR DESCRIPTION
**Pull request by DamienCassou**
Having fun trying to package Gnome 3.20.

Issue: #14289

cc @lethalman 

---

To be done:
- [x] Aisleriot doesn't build ([upstream bug](https://bugzilla.gnome.org/show_bug.cgi?id=764305), but this is not blocking for us) : fixed by making a libcanberra version depending on `gtk+-3.20`
- [ ] Fix commits with `(todo)` in their message
- [ ] Enable Google authentication in evolution-data-server
- [ ] Make sure  the `nixos-rebuild` output depends on latest libraries:
  - no gtk 3.18
  - no gnome 3.18
  - latest clutter

### CODE REVIEW

**Comment by jgillich** [comment]
:+1:  I'll try this soon. Any issues you've been running into so far?

**Comment by DamienCassou** [comment]
```
/tmp/nix-build-aisleriot-3.20.1.drv-0/aisleriot-3.20.1/src/board-noclutter.c:3423: undefined reference to `gtk_widget_class_set_css_name'
collect2: error: ld returned 1 exit status
Makefile:606: recipe for target 'sol' failed
make[4]: *** [sol] Error 1
make[4]: Leaving directory '/tmp/nix-build-aisleriot-3.20.1.drv-0/aisleriot-3.20.1/src'
Makefile:896: recipe for target 'all-recursive' failed
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory '/tmp/nix-build-aisleriot-3.20.1.drv-0/aisleriot-3.20.1/src'
Makefile:519: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/tmp/nix-build-aisleriot-3.20.1.drv-0/aisleriot-3.20.1/src'
Makefile:497: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/tmp/nix-build-aisleriot-3.20.1.drv-0/aisleriot-3.20.1'
Makefile:428: recipe for target 'all' failed
make: *** [all] Error 2
builder for ‘/nix/store/jc6x2696wd3li1al1mardsz6826ih182-aisleriot-3.20.1.drv’ failed with exit code 2
error: build of ‘/nix/store/jc6x2696wd3li1al1mardsz6826ih182-aisleriot-3.20.1.drv’ failed
```

File: pkgs/desktops/gnome-3/3.20/default.nix [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+      self = self_ // overrides;
+      self_ = with self; {
+
+  overridePackages = f:
+    let newself = pkgsFun (f newself self);
+    in newself;
+
+  callPackage = pkgs.newScope self;
+
+  version = "3.18";
```
Comment (lethalman) [inline pkgs/desktops/gnome-3/3.20/default.nix]: This is wrong.

**Comment by lethalman** [comment]
@DamienCassou it's probable that aisleriot is pulling gtk 3.18 instead of gtk 3.20 due to some dependency relying on gtk 3.18. You can find that with `nix-store -qR` of the aisleriot .drv.

**Comment by lethalman** [comment]
Also, like for aisleriot, we must be sure that the nixos-rebuild output does not depend on gtk 3.18 or any of the gnome 3.18 dependencies.

**Comment by DamienCassou** [comment]
@lethalman what do you think about adding that to `all-packages.nix`:

```diff
 
+  webkitgtk_gtk3_20 = callPackage ../development/libraries/webkitgtk {
+    harfbuzz = harfbuzz-icu;
+    gst-plugins-base = gst_all_1.gst-plugins-base;
+    gtk3 = gtk3_20;
+  };
+
```

**Comment by lethalman** [comment]
@DamienCassou looks ok, what about using `.override` instead?

**Comment by vcunat** [comment]
Do you expect some reason to keep both gtk+ 3.18 and 3.20? It would seem much easier to just have one `gtk3` and try to fix (potential) problems with apps using the newer version.

**Comment by lethalman** [comment]
@vcunat no there's no reason to keep both versions. The last time it was a huge regression in styling, but I hope it's no more the case. That is, gnome 3.18 should work fine with gtk 3.20, but it must be tested... I wouldn't bet on it.

**Comment by vcunat** [comment]
Hmm, the relevant section of [readme](https://git.gnome.org/browse/gtk+/plain/README.in?id=3.20.0) suggests there might be some compatibility issues, e.g. require some CSS theming changes (again). I can't see how serious it might be, but I think we can chance this on master/staging.

**Comment by DamienCassou** [comment]
@lethalman can you tell me more about that?
@vcunat I thought it would be a simple first step to do it like this (this is the first time I do something as large as this task). I can do it your way if you think it's easier.

**Comment by lethalman** [comment]
This is quite scary "The way theming works in GTK+ has been reworked fundamentally". Btw we did this transition in the past, what about doing it again @vcunat ? We'll do out best to drop gnome 3.18.

**Comment by vcunat** [comment]
Hmm, I see there's no mention of theming rework in previous release notes, but I hope this time won't be worse than the previous time.

**Comment by lethalman** [comment]
@DamienCassou did you happen to test gnome 3.18 with gtk 3.20 ?

**Comment by DamienCassou** [comment]
@lethalman no, I'm focusing on packagnig gnome 3.20 right now.

**Comment by lethalman** [comment]
@DamienCassou ok so I'll try to test gtk 3.20 on gnome 3.18 in the next days.

**Comment by DamienCassou** [comment]
@lethalman is it normal that the number of commits keeps on increasing? I don't see the end :-) (and webkit is getting built for the 5th time).

@k0ral this pull request contains `webkitgtk_gtk3_20` with webkit 2.12.

**Comment by vcunat** [comment]
Do you feel a Hydra jobset would help?

**Comment by DamienCassou** [comment]
@vcunat I've never worked in coordination with Hydra, I don't know how it would help. Maybe it's time to try :-).

**Comment by vcunat** [comment]
I think it will be beset if you push this WIP as gnome-3.20 branch in the *official* nixpkgs repository and I will create a jobset for that, OK? (I trust you do have push access.)

**Comment by DamienCassou** [comment]
@vcunat done. What does it change for me? What do I have to do?

**Comment by vcunat** [comment]
I created http://hydra.nixos.org/jobset/nixos/gnome-3.20. It gets lower priority than the usual branches so it will be slower, but we should see build breakages in there and get binaries into the official cache immediately.

**Comment by lethalman** [comment]
gtk 3.20.2 works well with gnome 3.18. We can drop gtk 3.18 altogether instead of keeping a separate version.

**Comment by DamienCassou** [comment]
@vcunat thank you. I will have a look.

@lethalman I'm preparing a rebase of this branch based on your tests. Thanks. Should I do the same for webkitgtk and just update its version?

**Comment by lethalman** [comment]
@DamienCassou yes, webkitgtk will just depend upon the new gtk.

**Comment by DamienCassou** [comment]
@vcunat sorry, I have no idea how to read http://hydra.nixos.org/jobset/nixos/gnome-3.20. Is there some documentation around?

**Comment by vcunat** [comment]
Reading comparisons to master will probably be most useful. http://hydra.nixos.org/eval/1251962?compare=trunk-combined There have been some fixes in there since, leading to showing incorrect "new failures", but probably not too many ATM.

**Comment by DamienCassou** [comment]
Move PR to #14366 because of changing remote (from https://github.com/DamienCassou/nixpkgs to https://github.com/NixOS/nixpkgs).

**Pull request closed without merging**